doc/design/scope.md

```diff
@@ -0,0 +1,58 @@
+# Scope
+
+### Define
+
+Scope is a context to manage Variables. It mainly contains a map from Variable name to Variable. Net will get and update variable throw scope.
+
+```cpp
+class Variable;
+using VariablePtr = std::shared_ptr<Variable>;
```

Comment:
When using `unique_ptr` and return a raw pointer, users must not hold this pointer inside private member or any global variable. Because when `scope` is destroyed, all pointers get before are invalid.

I think to use `shared_ptr` inside a `Scope`, but return a `weak_ptr` when `GetVariable` can resolve this issue. But this code is a little bit complicated.

Maybe we should have an agreement. We cannot store any variable pointer in our code.